### PR TITLE
BLE server controller

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

A BLE controller using the BLE server.

This exposes all sensors, text_sensors, binary_sensors, and switches.
Switches can also be turned on/off via writing to the characteristic.

This is built on top of #1807 and needs that merged first.

Some ideas came from https://github.com/wifwucite/esphome-ble-controller

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>
  
# Test Environment

- [ ] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.
Very important to fill if no issue linked

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
Here is a true diff for this PR: https://github.com/jesserockz/esphome/compare/ble-wifi-provisioning...esphome:ble-server-controller

File: esphome/components/esp32_ble_controller/__init__.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            "The esp32_ble_controller log level {} must be less severe than the "
+            "global log level {}.".format(level, global_level)
+        )
+
+
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.declare_id(BLEController),
+        cv.GenerateID(CONF_BLE_SERVER_ID): cv.use_id(esp32_ble_server.BLEServer),
+        cv.Optional(CONF_LOG_LEVEL): cv.All(
```
Comment (OttoWinter) [inline esphome/components/esp32_ble_controller/__init__.py:44]: This should be set using `logger.logs` - or is there another reason for having this option?

**Review by OttoWinter — COMMENTED** [review]
Looks good after a quick look.

What are the security properties of this? Does _any_ device in the vicinity have the ability to connect+control the device, or does it need to be bonded first? If so, what is the pairing process like (usually with BLE you need to confirm a PIN).

**Comment by marklister** [comment]
I wasn't able to get writing to a switch working. When I transmit a byte array of 0x0000 or 0x0001 or 0x0100 it appears to do an array append. Toggling the switch via other means works as expected.

File: esphome/components/esp32_ble_controller/ble_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+  {
+    auto switches = App.get_switches();
+    if (!switches.empty()) {
+      this->switch_service_ = global_ble_server->create_service(SWITCH_SERVICE_UUID);
+      for (auto *obj : switches) {
+        std::string uuid = std::string(SWITCH_SERVICE_UUID).substr(0, 28);
+        uuid += uint32_to_string(obj->get_object_id_hash());
+        BLECharacteristic *characteristic = this->switch_service_->create_characteristic(
+            uuid,
+            BLECharacteristic::PROPERTY_READ | BLECharacteristic::PROPERTY_NOTIFY | BLECharacteristic::PROPERTY_WRITE);
```
Comment (marklister) [inline esphome/components/esp32_ble_controller/ble_controller.cpp:118]: `  BLECharacteristic::PROPERTY_READ | BLECharacteristic::PROPERTY_NOTIFY | BLECharacteristic::PROPERTY_WRITE | BLECharacteristic::PROPERTY_WRITE_NR);
`

seems to allow writing to a switch.

**Comment by ctrix** [comment]
Hello all, i think that this `ble-controller` addon would be extremely useful and i would like to push for the inclusion in the mainline code.
Is there something i could help with?

**Pull request closed without merging**